test/prober/await.go

```diff
@@ -0,0 +1,155 @@
+ */
+
+package prober
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"knative.dev/eventing/test/lib"
+	"knative.dev/eventing/test/lib/duck"
+	"knative.dev/eventing/test/lib/resources"
+	"sync"
+	"time"
+)
+
+func (p *prober) waitForKServiceReady(name, namespace string) error {
+	meta := resources.NewMetaResource(name, namespace, &servingType)
+	return duck.WaitForResourceReady(p.client.Dynamic, meta)
+}
+
+func (p *prober) waitForKServiceScale(name, namespace string, satisfyScale func(*int32) bool) error {
+	metricName := fmt.Sprintf("waitForKServiceScale/%s/%s", namespace, name)
+	_, span := trace.StartSpan(context.Background(), metricName)
+	defer span.End()
+
+	return wait.PollImmediate(duck.Interval, duck.Timeout, func() (bool, error) {
+		serving := p.client.Dynamic.Resource(servicesCR).Namespace(namespace)
+		unstruct, err := serving.Get(name, metav1.GetOptions{})
+		return p.isScaledTo(satisfyScale, unstruct, namespace, err)
+	})
+}
+
+func (p *prober) isScaledTo(satisfyScale func(*int32) bool, un *unstructured.Unstructured, namespace string, err error) (bool, error) {
+	if k8serrors.IsNotFound(err) {
+		// Return false as we are not done yet.
+		// We swallow the error to keep on polling.
+		// It should only happen if we wait for the auto-created resources, like default Broker.
+		return false, nil
+	} else if err != nil {
+		// Return error to stop the polling.
+		return false, err
+	}
+
+	content := un.UnstructuredContent()
+	maybeStatus, ok := content["status"]
+	if !ok {
+		return false, nil
+	}
+	status := maybeStatus.(map[string]interface{})
+	maybeTraffic, ok := status["traffic"]
+	if !ok {
+		return false, nil
+	}
+	traffic := maybeTraffic.([]interface{})
+	if len(traffic) > 1 {
+		return false, fmt.Errorf("traffic shouldn't be split to more then 1 revision: %v", traffic)
+	}
+	if len(traffic) == 0 {
+		// continue to wait
+		return false, nil
+	}
+	firstTraffic := traffic[0].(map[string]interface{})
+	revisionName := firstTraffic["revisionName"].(string)
+	deploymentName := fmt.Sprintf("%s-deployment", revisionName)
+
+	var dep *appsv1.Deployment
+	dep, err = p.client.Kube.Kube.AppsV1().Deployments(namespace).
+		Get(deploymentName, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		// Return false as we are not done yet.
+		return false, nil
+	} else if err != nil {
+		// Return error to stop the polling.
+		return false, err
+	}
+	return satisfyScale(dep.Spec.Replicas), nil
+}
+
+func (p *prober) waitForTriggerReady(name, namespace string) error {
+	meta := resources.NewMetaResource(name, namespace, lib.TriggerTypeMeta)
+	return duck.WaitForResourceReady(p.client.Dynamic, meta)
+}
+
+func (p *prober) waitForPodReady(name, namespace string) error {
```

Comment:
We have a `WaitForPodRunning` function in https://github.com/knative/pkg/blob/f609dc07e3a16e3278c1947c527a93269ed13957/test/kube_checks.go#L154

Reply:
Okay. I'll use that.